### PR TITLE
Updates for GraalVM

### PR DESCRIPTION
**Pull request by pitr-ch**

### CODE REVIEW

File: bin/truffleruby [hunk 1]
```diff
@@ -120,7 +138,28 @@ if [ -n "$CP" ]; then
...
+    full_command=(
+        $JAVACMD # no " to let options expand to multiple array elements
+        $JAVA_OPTS
+        "${java_args[@]}"
+        com.oracle.graalvm.Main
+        org.truffleruby.Main
+        "-Xlauncher=$root/bin/truffleruby"
+        "${ruby_args[@]}"
+    )
+fi
```
Comment (eregon) [inline bin/truffleruby]: There is almost no shared code (just `-J-cmd` and parsing simple options) beside the symlink resolving which needs to happen in all GraalVM launchers anyway.
I think it's not worth the extra complexity. One could make a shared bash file for shared functions if duplication of the flags parsing is so important.

**Comment by chrisseaton** [comment]
It seems ok, but this is really fragile code for many reasons. Ok to merge I think, but we should integrate into GraalVM quickly and ensure all tests everywhere pass, and then do some manual testing of the tarball.

File: bin/truffleruby (hunk 1 above)
Comment (eregon) [inline bin/truffleruby]: It becomes really hard to follow https://github.com/pitr-ch/truffleruby/blob/1fe6464c828680da1bcb87eb933fbe16a40ad575/bin/truffleruby compared to https://github.com/graalvm/truffleruby/blob/master/bin/truffleruby.
Given this is bash, it's more error-prone and harder to follow any logic.
I ask to keep them split as they share close to nothing.
Moreover, GraalVM-specific stuff (e.g.: `-noverify -polyglot` and paths) should live in GraalVM, not our launcher. It also makes it easier to compared to other launguage launchers.

File: bin/truffleruby [hunk 2]
```diff
@@ -120,7 +138,28 @@ if [ -n "$CP" ]; then
...
+        "-Xbootclasspath/a:$boot_cp_jar"
+        "${java_args[@]}"
+        org.truffleruby.Main
+        "-Xhome=$root"
+        "-Xlauncher=$root/bin/truffleruby"
+        "${ruby_args[@]}"
+    )
+else
+    full_command=(
+        $JAVACMD # no " to let options expand to multiple array elements
```
Comment (eregon) [inline bin/truffleruby]: This could be simplified by inlining the definition of JAVACMD above.

File: bin/truffleruby (hunk 1 above)
Comment (chrisseaton) [inline bin/truffleruby]: Our repo launcher and the GraalVM launcher have been different in the past and that caused some problems, so I hope having one will reduce that. But it is a bit more complicated.

File: bin/truffleruby [hunk 3]
```diff
@@ -120,7 +138,28 @@ if [ -n "$CP" ]; then
...
+if [ $on_graalvm = false ]; then
+    full_command=(
+        "$JAVACMD"
+        $JAVA_OPTS
+        "-Xbootclasspath/a:$boot_cp_jar"
+        "${java_args[@]}"
+        org.truffleruby.Main
+        "-Xhome=$root"
+        "-Xlauncher=$root/bin/truffleruby"
+        "${ruby_args[@]}"
```
Comment (eregon) [inline bin/truffleruby]: These last 4 arguments could be shared.

File: bin/truffleruby (hunk 2 above)
Comment (pitr-ch) [inline bin/truffleruby]: done

File: bin/truffleruby [hunk 4]
```diff
@@ -16,18 +16,6 @@ if [ -n "$RUBY_BIN" ]; then
...
-    uname_str=$(uname)
-    if [ "$uname_str" = 'Linux' ] || [ "$uname_str" = 'SunOS' ]; then
-        stat -c '%Y' "$1" 2>/dev/null || echo 0
-    elif [ "$uname_str" = 'Darwin' ]; then
-        stat -f '%m' "$1" 2>/dev/null || echo 0
-    else
-        echo "unknown platform $uname_str" 1>&2
-        exit 1
-    fi
-}
```
Comment (eregon) [inline bin/truffleruby]: I think it's best to keep functions at the top, and not conditionally define them, particularly in the middle of the logic.

File: bin/truffleruby (hunk 1 above)
Comment (pitr-ch) [inline bin/truffleruby]: I'd rather have one file then more, we've already forgot to carry over a fix. Anyway there is not much time we can revisit.

File: bin/truffleruby (hunk 3 above)
Comment (pitr-ch) [inline bin/truffleruby]: Not without changes, there is an extra -Xhome=$root

Comment (eregon) [inline bin/truffleruby]: That's the correct value, isn't it?

File: bin/truffleruby [hunk 5]
```diff
@@ -39,44 +27,74 @@ while [ -h "$SELF_PATH" ]; do
...
-    JAVACMD='java'
-  else
-    JAVACMD="$JAVA_HOME/bin/java"
-  fi
+root_parent=$(dirname "$root")
+if [ $(basename "$root") = ruby ] && [ $(basename "$root_parent") = language ]; then
+    on_graalvm=true
+else
+    on_graalvm=false
+fi
```
Comment (eregon) [inline bin/truffleruby]: One way to make this less hacky would be to have `graalvm/bin/ruby` define an extra variable `$GRAALVM` or so, and then `source` this script. It does mean duplicating the symlink expansion though.

Comment (pitr-ch) [inline bin/truffleruby]: It has to work for calling `graalvm/language/ruby/bin/ruby`as well though.

File: bin/truffleruby (hunk 4 above)
Comment (pitr-ch) [inline bin/truffleruby]: The script has 1 function and it is only used in the block so I've moved it next to its call-sites. If the launcher grows in complexity (which I hope will not) then yeah it would be good to group the functions at the top. Keeping it here for now.

File: bin/truffleruby (hunk 3 above)
Comment (pitr-ch) [inline bin/truffleruby]: I am not sure, I don't want to investigate it right now.

Comment (eregon) [inline bin/truffleruby]: Or you could trust me, I know because I implemented the magic lookup from `ruby.jar`.

File: bin/truffleruby (hunk 5 above)
Comment (eregon) [inline bin/truffleruby]: Good point

File: bin/truffleruby (hunk 4 above)
Comment (eregon) [inline bin/truffleruby]: There is no reason to move. Conditional function definition is always a bad idea..

File: bin/truffleruby (hunk 3 above)
Comment (eregon) [inline bin/truffleruby]: And the whole purpose of this is to make language/ruby look like a normal Ruby home, so it only makes sense.

Comment (pitr-ch) [inline bin/truffleruby]: In that case I can simplify, I thought you are unsure as well. My understanding of "...  isn't it?" might be bad.

Comment (eregon) [inline bin/truffleruby]: The `isn't it?` was meant as "that's what I believe" and also asking to check by yourself so you get more confidence about it.
Look in `graalvm-0.20/language/ruby`, that's the home, where there is `lib`, `doc` and soon `bin`.
